### PR TITLE
Runner Tcp support

### PR DESCRIPTION
**Pull request by sharadregoti**
1) Runner now supports tcp base traffic splitting

### CODE REVIEW

File: runner/model/routing.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ type Route struct {
 	ID             string        `json:"id" yaml:"id"`
 	RequestRetries int32         `json:"requestRetries" yaml:"requestRetries"`
 	RequestTimeout int64         `json:"requestTimeout" yaml:"requestTimeout"`
+	Protocol       Protocol      `json:"protocol" yaml:"protocol"`
```
Comment (YourTechBud) [inline runner/model/routing.go:26]: This will come in source

File: runner/utils/driver/istio/helpers.go [hunk 2]
```diff
@@ -173,6 +173,9 @@ func prepareVirtualServiceHTTPRoutes(ctx context.Context, projectID, serviceID s
 	var httpRoutes []*networkingv1alpha3.HTTPRoute
 
 	for _, route := range routes {
+		if route.Protocol != "" && route.Protocol != model.HTTP {
+			continue
+		}
 		// Check if the port provided is correct
```
Comment (YourTechBud) [inline runner/utils/driver/istio/helpers.go:179]: Add a space before this line

File: runner/utils/driver/istio/helpers.go [hunk 3]
```diff
@@ -173,6 +173,9 @@ func prepareVirtualServiceHTTPRoutes(ctx context.Context, projectID, serviceID s
 	var httpRoutes []*networkingv1alpha3.HTTPRoute
 
 	for _, route := range routes {
+		if route.Protocol != "" && route.Protocol != model.HTTP {
```
Comment (YourTechBud) [inline runner/utils/driver/istio/helpers.go:176]: Its time we start commenting why we are doing certain things

File: runner/utils/driver/istio/helpers.go [hunk 4]
```diff
@@ -257,6 +260,76 @@ func prepareVirtualServiceHTTPRoutes(ctx context.Context, projectID, serviceID s
...
 }
 
+func prepareVirtualServiceTCPRoutes(ctx context.Context, projectID, serviceID string, services map[string]model.AutoScaleConfig, routes model.Routes, proxyPort uint32) ([]*networkingv1alpha3.TCPRoute, error) {
+	var tcpRoutes []*networkingv1alpha3.TCPRoute
+
+	for _, route := range routes {
+		if route.Protocol != model.TCP {
+			continue
+		}
+		// Check if the port provided is correct
```
Comment (YourTechBud) [inline runner/utils/driver/istio/helpers.go:270]: Add space before this line

File: runner/utils/driver/istio/helpers.go [hunk 5]
```diff
@@ -257,6 +260,76 @@ func prepareVirtualServiceHTTPRoutes(ctx context.Context, projectID, serviceID s
 	return httpRoutes, nil
 }
 
+func prepareVirtualServiceTCPRoutes(ctx context.Context, projectID, serviceID string, services map[string]model.AutoScaleConfig, routes model.Routes, proxyPort uint32) ([]*networkingv1alpha3.TCPRoute, error) {
+	var tcpRoutes []*networkingv1alpha3.TCPRoute
+
+	for _, route := range routes {
+		if route.Protocol != model.TCP {
```
Comment (YourTechBud) [inline runner/utils/driver/istio/helpers.go:267]: Add comment on why we are continuing

File: runner/utils/driver/istio/helpers.go [hunk 6]
```diff
@@ -257,6 +260,76 @@ func prepareVirtualServiceHTTPRoutes(ctx context.Context, projectID, serviceID s
...
+
+				// Prepare variables
+				destHost := getInternalServiceDomain(projectID, serviceID, target.Version)
+				destPort := uint32(target.Port)
+
+				// Redirect traffic to runner when no of replicas is equal to zero. The runner proxy will scale up the service to service incoming requests.
+				if versionScaleConfig.MinReplicas == 0 {
+					destHost = "runner.space-cloud.svc.cluster.local"
+					destPort = proxyPort
+				}
```
Comment (YourTechBud) [inline runner/utils/driver/istio/helpers.go:299]: Runner can only handle http requests. Always direct to service. You dont really need versionScaleConfig for tcp.

File: runner/utils/driver/istio/helpers.go [hunk 7]
```diff
@@ -828,10 +901,18 @@ func (i *Istio) generateVirtualServiceBasedOnRoutes(ctx context.Context, project
...
+	if err != nil {
+		return nil, err
+	}
+
 	// Create a prevVirtualService if its nil to prevent panic
 	if prevVirtualService == nil {
 		prevVirtualService = new(v1alpha3.VirtualService)
 	}
+	if len(tcpRoutes) == 0 {
+		tcpRoutes = prevVirtualService.Spec.Tcp
```
Comment (YourTechBud) [inline runner/utils/driver/istio/helpers.go:914]: This is a bug. No tcp routes means we don't need tcp routes anymore.

File: runner/utils/driver/istio/helpers.go [hunk 8]
```diff
@@ -828,10 +901,18 @@ func (i *Istio) generateVirtualServiceBasedOnRoutes(ctx context.Context, project
...
 
+	tcpRoutes, err := prepareVirtualServiceTCPRoutes(ctx, projectID, serviceID, scaleConfig, routes, i.config.ProxyPort)
+	if err != nil {
+		return nil, err
+	}
+
 	// Create a prevVirtualService if its nil to prevent panic
 	if prevVirtualService == nil {
 		prevVirtualService = new(v1alpha3.VirtualService)
 	}
```
Comment (YourTechBud) [inline runner/utils/driver/istio/helpers.go:912]: I dont think we need the prevVirtualService at all.

**Review by YourTechBud — CHANGES_REQUESTED** [review]
Update the postman config for service routes. Add the protocol, requestRetries & request timeout field

File: runner/utils/driver/istio/helpers_test.go [hunk 9]
```diff
@@ -48,10 +48,10 @@ func Test_prepareVirtualServiceTCPRoutes(t *testing.T) {
 				},
 				routes: []*model.Route{
 					{
-						ID:       "414a27e4-e3d2-4003-9567-0ad949e25c8e",
-						Protocol: model.TCP,
+						ID: "414a27e4-e3d2-4003-9567-0ad949e25c8e",
```
Comment (YourTechBud) [inline runner/utils/driver/istio/helpers_test.go:51]: Isn't the id for service id?

File: postman/Space Up.postman_collection.json [hunk 10]
```diff
@@ -3706,6 +3706,11 @@
       "key": "gateway_base_url",
       "value": "http://localhost:4122"
     },
+    {
+      "id": "03da5b38-e050-44cc-b00a-cef12e3a7812",
+      "key": "deployed_service_idl",
```
Comment (YourTechBud) [inline postman/Space Up.postman_collection.json:3711]: Should be this right?
```suggestion
      "key": "deployed_service_id",
```

File: runner/utils/driver/istio/get.go [hunk 11]
```diff
@@ -307,7 +307,44 @@ func (i *Istio) GetServiceRoutes(ctx context.Context, projectID string) (map[str
 			}
 
 			// Set the route
-			routes[i] = &model.Route{ID: serviceID, Source: model.RouteSource{Port: int32(route.Match[0].Port)}, Targets: targets}
+			routes[i] = &model.Route{ID: serviceID, Source: model.RouteSource{Port: int32(route.Match[0].Port), Protocol: model.HTTP}, Targets: targets}
```
Comment (YourTechBud) [inline runner/utils/driver/istio/get.go:310]: Request retries and timeout?

File: runner/utils/driver/istio/get_test.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	tests := []struct {
+		name                      string
+		args                      args
+		virtualServiceToBeCreated *v1alpha3.VirtualService
+		want                      map[string]model.Routes
+		wantErr                   bool
+	}{
+		{
+			name: "Get TCP Routes with internal & external targets",
```
Comment (YourTechBud) [inline runner/utils/driver/istio/get_test.go:31]: Can we add test case for http too?

File: runner/utils/driver/istio/names.go [hunk 13]
```diff
@@ -31,6 +32,29 @@ func getInternalServiceDomain(projectID, serviceID, version string) string {
 	return fmt.Sprintf("%s.%s.svc.cluster.local", getInternalServiceName(serviceID, version), projectID)
 }
 
+func checkIfInternalServiceDomain(projectID, serviceID, internalServiceDomain string) bool {
+	arr := strings.Split(internalServiceDomain, ".")
+	if strings.HasSuffix(internalServiceDomain, "svc.cluster.local") {
+		if len(arr) == 5 && arr[1] == projectID && strings.HasPrefix(arr[0], serviceID) {
```
Comment (YourTechBud) [inline runner/utils/driver/istio/names.go:38]: Dont do prefix match for service id. Split by `-` and do exact match for service.

File: runner/utils/driver/istio/names.go [hunk 14]
```diff
@@ -31,6 +32,29 @@ func getInternalServiceDomain(projectID, serviceID, version string) string {
...
+	arr := strings.Split(s, ".")
+	if len(arr) < 5 {
+		return "", "", ""
+	}
+	t := strings.Split(arr[0], "-")
+	if len(t) < 2 {
+		return "", "", ""
+	}
+
+	return arr[1], strings.Join(t[:len(t)-3], ""), t[len(t)-2]
```
Comment (YourTechBud) [inline runner/utils/driver/istio/names.go:55]:
```suggestion
	return arr[1], strings.Join(t[:len(t)-3], ""), t[len(t)-1]
```

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: runner/utils/driver/istio/get_test.go (hunk 12 above)
Comment (sharadregoti) [reply runner/utils/driver/istio/get_test.go:31]: YES

File: runner/utils/driver/istio/names.go (hunk 14 above)
Comment (sharadregoti) [reply runner/utils/driver/istio/names.go:55]: This will be 2 only, the last element is `internal`

File: runner/utils/driver/istio/names.go [hunk 15]
```diff
@@ -27,10 +28,42 @@ func getInternalServiceName(serviceID, version string) string {
...
+	}
+	return strings.Join(arr[:len(arr)-2], "-"), arr[len(arr)-2]
+}
+
 func getInternalServiceDomain(projectID, serviceID, version string) string {
 	return fmt.Sprintf("%s.%s.svc.cluster.local", getInternalServiceName(serviceID, version), projectID)
 }
 
+func checkIfInternalServiceDomain(projectID, serviceID, internalServiceDomain string) bool {
+	arr := strings.Split(internalServiceDomain, ".")
```
Comment (YourTechBud) [inline runner/utils/driver/istio/names.go:44]: Take this inside the if

File: runner/utils/driver/istio/names.go [hunk 16]
```diff
@@ -27,10 +28,42 @@ func getInternalServiceName(serviceID, version string) string {
...
+	if strings.HasSuffix(internalServiceDomain, "svc.cluster.local") {
+		serID, _ := splitInternalServiceName(arr[0])
+		if len(arr) == 5 && arr[1] == projectID && serID == serviceID {
+			return true
+		}
+	}
+	return false
+}
+
+func splitInternalServiceDomain(s string) (projectID, serviceID, version string) {
```
Comment (YourTechBud) [inline runner/utils/driver/istio/names.go:54]: I don't think we need this function. Make the check function return everything along with thr Boolean

**Review by YourTechBud — CHANGES_REQUESTED** [review]

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**